gobblin-restli/gobblin-throttling-service/gobblin-throttling-service-server/src/main/java/org/apache/gobblin/restli/throttling/DynamicTokenBucket.java

```diff
@@ -63,39 +74,56 @@
    * @param minPermits the minimum number of tokens useful for the calling process. If this many tokens cannot be acquired,
    *                   the method will return 0 instead,
    * @param timeoutMillis the maximum wait the calling process is willing to wait for tokens.
-   * @return the number of allocated tokens.
+   * @return a {@link PermitsAndDelay} for the allocated permits.
    */
-  public long getPermits(long requestedPermits, long minPermits, long timeoutMillis) {
-
+  public PermitsAndDelay getPermitsAndDelay(long requestedPermits, long minPermits, long timeoutMillis) {
     try {
       long storedTokens = this.tokenBucket.getStoredTokens();
 
       long eagerTokens = storedTokens / 2;
       if (eagerTokens > requestedPermits && this.tokenBucket.getTokens(eagerTokens, 0, TimeUnit.MILLISECONDS)) {
-        return eagerTokens;
+        return new PermitsAndDelay(eagerTokens, 0, true);
       }
 
       long millisToSatisfyMinPermits = (long) (minPermits / this.tokenBucket.getTokensPerMilli());
       if (millisToSatisfyMinPermits > timeoutMillis) {
-        return 0;
+        return new PermitsAndDelay(0, 0, false);
       }
       long allowedTimeout = Math.min(millisToSatisfyMinPermits + this.baseTimeout, timeoutMillis);
 
       while (requestedPermits > minPermits) {
-        if (this.tokenBucket.getTokens(requestedPermits, allowedTimeout, TimeUnit.MILLISECONDS)) {
-          return requestedPermits;
+        long wait = this.tokenBucket.tryReserveTokens(requestedPermits, allowedTimeout);
+        if (wait >= 0) {
+          return new PermitsAndDelay(requestedPermits, wait, true);
         }
         requestedPermits /= 2;
       }
 
-      if (this.tokenBucket.getTokens(minPermits, allowedTimeout, TimeUnit.MILLISECONDS)) {
-        return minPermits;
+      long wait = this.tokenBucket.tryReserveTokens(minPermits, allowedTimeout);
+      if (wait >= 0) {
+        return new PermitsAndDelay(requestedPermits, wait, true);
       }
+
     } catch (InterruptedException ie) {
       // Fallback to returning 0
     }
 
-    return 0;
+    return new PermitsAndDelay(0, 0, true);
```

Comment:
Why `possibleToSatisfy ` is still true in this case?

Reply:
Impossible to satisfy means the policy cannot ever grant that many permits. In this case, it might have been that there is just too much contention at the moment, so the permits cannot be granted just now but may be later.